types/jest/index.d.ts

```diff
@@ -121,7 +121,7 @@ declare namespace jest {
     /**
      * Creates a mock function. Optionally takes a mock implementation.
      */
-    function fn<T, Y extends any[]>(implementation: (...args: Y) => T): Mock<T, Y>;
+    function fn<F extends (...args: any[]) => any>(implementation?: F): Mock<ReturnType<F>, ArgsType<F>>;
```

Comment:
You can't change generics without a breaking change. It has to stay the same.

Before I introduced type inference in jest, it was:
```typescript
function fn<T>(implementation?: (...args: any[]) => any): Mock<T>;
```
I added a second generic for arguments
```typescript
function fn<T, Y extends any[]>(implementation: (...args: Y) => T): Mock<T, Y>;
```
The correct fix would be to set implementation as optional, without touching generics.
```typescript
function fn<T, Y extends any[]>(implementation?: (...args: Y) => T): Mock<T, Y>;
```
And add a test:
```
interface TestFn {
    test(x: number): string
}
// $ExpectType Mock<string, [number]>
const mock12 = jest.fn<ReturnType<TestFn['test']>, ArgsType<TestFn['test']>>();
```

Reply:
To be honest, I much prefer the previous template, that's why I re-introduced it. In fact I believe that your upgrade has introduced unnecessary breaking changes (as compared to 23.x) that are actually regressions because they reduce typing in many cases and introduce potential for inconsistency. Example: I believe that something like jest.fn<number> or jest.fn<..., [never]> should NOT be valid. The same is true for other types like Mock which I didn't touch. But don't get me wrong: Introducing typed arguments was of course a good thing.

What was your rationale to do it like this? Do you agree in principle that my solution would have been easier to upgrade to for most 23.x users, introduces less redundancy and potential for error? Or am I just unaware of some subtlety or less obvious use case?

Maybe you can explain this a little for me and other who happen to stumble upon this?

If we agree that this solution is objectively better, though, then of course you're right that it is still formally another breaking change. :-( As the types packages are not semversioned there might be a way to change this anyways before most users try to upgrade and stumble upon this incompatibility? In practice this might be less of a breaking change for many as your code is only out since a few days.

Of course I'm just guessing and it's for you and the maintainers to decide...

Reply:
> Example: I believe that something like `jest.fn<number>` or `jest.fn<..., [never]>` should NOT be valid.

Had a long discussion about default values in my first PR #31094 . There was a lot of pushback about removing defaults and introducing breaking change in existing generics. I would like to simplify the generics like you too. 

Passing the function as the generic should've been the way when they first did it, but it wasn't. I agree that we should change it in the future, I just don't think we should do it for a bug in 1 declaration. If we change it, it has to be consistent across the api (Mock, MockInstance, SpyInstance, etc.)

Reply:
> Example: I believe that something like `jest.fn<number>` or `jest.fn<..., [never]>` should NOT be valid.

Had a long discussion about default values in my first PR #31094 . There was a lot of pushback about removing defaults and introducing breaking change in existing generics. I would like to simplify the generics like you too. 

Passing the function as the generic should've been the way when they first did it, but it wasn't. I agree that we should change it in the future, I just don't think we should do it for a bug in 1 declaration. If we change it, it has to be consistent across the api (Mock, MockInstance, SpyInstance, etc.)

Reply:
Ok, I'll revert to the `<X, Y>` types and opened up an issue to track the `<X, Y>` -> `<F>` change, see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/32901. Stay tuned. :-)

Reply:
@jerico-dev would you be able to update your [Working sample implementation of the above mocking API](https://github.com/facebook/jest/issues/7832) to reflect this change, or otherwise update the link in the issue to point to your new PR, if that's required for it to work :)

Reply:
@G-Rath I'm sorry that my feature request still pointed to this issue. The actual change needed to make the feature request work is https://github.com/DefinitelyTyped/DefinitelyTyped/issues/32901 (see there). We initially had the two things mixed but then split them in two as one was easily acceptable to the maintainers and the other (the open one) isn't. As long as the maintainers do not show any interest in the change I won't update the PR.

Reply:
I updated https://github.com/facebook/jest/issues/7832 to point to the right issue/PR, thanks @G-Rath for your heads-up.